src/fides/service/dataset/dataset_config_service.py

```diff
@@ -0,0 +1,324 @@
+from fides.api.graph.traversal import Traversal
+from fides.api.graph.traversal_node import TraversalNode
+from fides.api.models.connectionconfig import ConnectionConfig
+from fides.api.models.datasetconfig import DatasetConfig
+from fides.api.models.policy import Policy
+from fides.api.models.privacy_request import (
+    PrivacyRequest,
+    PrivacyRequestSource,
+    PrivacyRequestStatus,
+)
+from fides.api.schemas.api import BulkUpdateFailed
+from fides.api.schemas.dataset import (
+    BulkPutDataset,
+    DatasetConfigCtlDataset,
+    ValidateDatasetResponse,
+)
+from fides.api.schemas.redis_cache import Identity, LabeledIdentity
+from fides.api.util.data_category import get_data_categories_from_db
+from fides.service.dataset.dataset_service import (
+    DatasetNotFoundException,
+    _get_ctl_dataset,
+)
+from fides.service.dataset.dataset_validator import DatasetValidator
+from fides.service.dataset.validation_steps.data_category import (
+    validate_data_categories_against_db,
+)
+from fides.service.dataset.validation_steps.traversal import TraversalValidationStep
+
+
+class DatasetFilter:
+    """
+    Filter that excludes nodes that are not part of the specified dataset.
+    This ensures that unreachable nodes from other datasets are not treated as errors.
+    """
+
+    def __init__(self, dataset_name: str):
+        self.dataset_name = dataset_name
+
+    def exclude_node(self, node: TraversalNode) -> bool:
+        """Returns True if the node is not part of the dataset"""
+        return node.address.dataset != self.dataset_name
+
+
+class DatasetConfigService:
+    def __init__(self, db: Session):
+        self.db = db
+
+    def create_or_update_dataset_config(
+        self,
+        connection_config: ConnectionConfig,
+        dataset: Union[DatasetConfigCtlDataset, FideslangDataset],
+    ) -> Tuple[Optional[FideslangDataset], Optional[BulkUpdateFailed]]:
+        """Create or update a single dataset"""
+        try:
+            if isinstance(dataset, DatasetConfigCtlDataset):
+                ctl_dataset = _get_ctl_dataset(self.db, dataset.ctl_dataset_fides_key)
+                dataset_to_validate = FideslangDataset.model_validate(ctl_dataset)
+                data_dict = {
+                    "connection_config_id": connection_config.id,
+                    "fides_key": dataset.fides_key,
+                    "ctl_dataset_id": ctl_dataset.id,
+                }
+            else:
+                dataset_to_validate = dataset
+                data_dict = {
+                    "connection_config_id": connection_config.id,
+                    "fides_key": dataset.fides_key,
+                    "dataset": dataset.model_dump(mode="json"),
+                }
+
+            # Validate dataset
+            DatasetValidator(
+                self.db,
+                dataset_to_validate,
+                connection_config,
+                skip_steps=[TraversalValidationStep],
+            ).validate()
```

Comment:
Centralized validation! We don't want to be strict and check traversability yet, we want to give the user the ability upload an unreachable dataset and work on it in the dataset tester UI